### PR TITLE
Labs 08

### PR DESCRIPTION
**Pull request by jorien-witjas**
Hi Gilad,

This pull request contains labs for 07 , 08 and 09.

### CODE REVIEW

File: python_fundamentals-master/04_conditionals_loops/04_01_divisible.py [hunk 1]
```diff
@@ -11,4 +11,6 @@
     print("your number is divisible by 3")
     print(f"the result is: ", number / 3)
 else:
-    print("your number is not divisible by 3")
\ No newline at end of file
+    print("your number is not divisible by 3")
+
+print('hello')
```
Comment (giladgressel) [inline python_fundamentals-master/04_conditionals_loops/04_01_divisible.py]: This should actually be print("jollo")

**Review by giladgressel — CHANGES_REQUESTED** [review]
please make that change

File: python_fundamentals-master/07_classes_objects_methods/07_01_car.py [hunk 2]
```diff
@@ -10,20 +10,25 @@
...
-print(smart.year)
-print(smart.speed_up(5))
-volvo = car("volvo", "2010", 260)
-print(volvo.model)
-print(volvo.year)
-print(volvo.speed_up(5))
+        self.speed = speed
+    def speed_up(self, speed_up):
+        self.speed_up = (self.speed + 5)
+        return f"max_speed is {self.speed_up}"
```
Comment (giladgressel) [inline python_fundamentals-master/07_classes_objects_methods/07_01_car.py]: Perfect, this actually modifies the `self.speed` attribute so the change would persist.

File: python_fundamentals-master/07_classes_objects_methods/07_02_shapes.py [hunk 3]
```diff
@@ -9,27 +9,25 @@
...
         self.radius = radius
-    def area(self, length, width):
-        self.area = length * width
-        return print(f"the area of the circle is {self.area}")
-    def circumference(self, r):
-        self.circumference() = 2 * 3,1415 * self.radius
-        return print(f"the circumference of the circle is {self.circumference()}")
\ No newline at end of file
+        self.area = (3.1415 * self.radius ** 2)
+        self.circumference = 2 * 3.1415 * self.radius
```
Comment (giladgressel) [inline python_fundamentals-master/07_classes_objects_methods/07_02_shapes.py]: So, you don't need to pass `area, circumference` into the `init` constructor.  You can just create them in `__init__` without actually passing them when the class is created.

You should think about what gets passed as arguments to `__init__` as "what information does the user have to specify?
They do not need to specify the area or circumference since this can be solely calculated from the radius.
Does this make sense?

Basically you can do
```
def __init__(self, radius ):
        self.radius = radius
        self.area = (3.1415 * self.radius ** 2)
        self.circumference = 2 * 3.1415 * self.radius
```

And it will work the same.

File: python_fundamentals-master/08_file_io/08_01_words_analysis.py [hunk 4]
```diff
@@ -7,3 +7,20 @@
 
 
 '''
+with open("words.txt" , "r") as f:
+    word_min = "hello"
```
Comment (giladgressel) [inline python_fundamentals-master/08_file_io/08_01_words_analysis.py]: why is `word_min` hardcoded to be `hello` ?
I think you were trying to find a large word to start with? In that case I would recommend you use `float('inf')` which is pythons positive infinity.  There is also a `float('-inf')` for negative infinity. These are commonly used for this purpose, setting limits.

image

File: python_fundamentals-master/08_file_io/08_01_words_analysis.py [hunk 5]
```diff
@@ -7,3 +7,20 @@
...
+    for word in f:
+        if len(word) <= len(word_min):
+            word_min = word
+            words_min.append(word_min.strip())
+        if len(word) >= len(word_max):
+            word_max = word
+        total_num_words += 1
+    print(words_min)
+    print(word_max)
+    print(total_num_words)
```
Comment (giladgressel) [inline python_fundamentals-master/08_file_io/08_01_words_analysis.py]: So this doesn't actually meet the objective right? Anytime a word is larger or smaller it will get added to the right list. but the list will contain _all_ the words which were found in the process.  You need to find the _smallest_ of those words.  This could be the last one added (but then how how do you check for ties?) 
I think this needs a bit more work.

File: python_fundamentals-master/09_exceptions/09_01_first_exception.py [hunk 6]
```diff
@@ -7,4 +7,10 @@
...
 
-'''
\ No newline at end of file
+'''
+
+try:
+    list_ = ["today it is sunny"]
+    print(list_[1])
+except IndexError:
+    print("first give me another item in the list")
```
Comment (giladgressel) [inline python_fundamentals-master/09_exceptions/09_01_first_exception.py]: Hah, cute

File: python_fundamentals-master/09_exceptions/09_05_check_for_ints.py [hunk 7]
```diff
@@ -6,3 +6,11 @@
...
 
 '''
+
+try:
+    num_1 = int(input("please give me a number: "))
+    num_2 = int(input("please give me another number: "))
+except ValueError:
+    print("please enter a number and no letters")
+else:
+    print("good job!")
```
Comment (giladgressel) [inline python_fundamentals-master/09_exceptions/09_05_check_for_ints.py]: This looks good, but it doesn't meet the requirements. It should keep asking until they do it.
So you need to do something about that 🤔

File: python_fundamentals-master/09_exceptions/09_06_check_files.py [hunk 8]
```diff
@@ -6,5 +6,16 @@
...
+        ints_read = ints.read()
+except FileNotFoundError:
+    print("the file does not exist, please create first")
+except ValueError:
+    print("there are strings in your file, I can't calculate with those. Please make sure your file only has integers")
+else:
+    num_1 = ints.readline(1)        #ValueError - file closed?
+    num_2 = ints.readline(2)
+    print(f"Sum of your integers is {num_1 + num_2}")
+
```
Comment (giladgressel) [inline python_fundamentals-master/09_exceptions/09_06_check_files.py]: nice

**Review by giladgressel — CHANGES_REQUESTED** [review]
looks really good!
good work.
A few small things only

File: python_fundamentals-master/07_classes_objects_methods/07_01_car.py (hunk 2 above)
Comment (jorien-witjas) [reply python_fundamentals-master/07_classes_objects_methods/07_01_car.py]: When I run the code, it gives me the following error:
    print(car_1.speed_up(210))
TypeError: 'int' object is not callable

File: python_fundamentals-master/07_classes_objects_methods/07_02_shapes.py (hunk 3 above)
Comment (jorien-witjas) [reply python_fundamentals-master/07_classes_objects_methods/07_02_shapes.py]: It does make sense, and I was struggling with this, because it didn't make sense to give it as an argument and it would calculate afterwards. For some reason I had errors before therefore I thought I had to specify all attributes before using them later on, but that is not needed. I changed it now and there are no errors anymore :) Good to know!

File: python_fundamentals-master/08_file_io/08_01_words_analysis.py (hunk 4 above)
Comment (jorien-witjas) [reply python_fundamentals-master/08_file_io/08_01_words_analysis.py]: Yes I was looking for a word longer than one to have a starting point, and couldn't think of a better option! ('inf') is great!

I tried to use float('inf') just now, but got this error:
    if len(word) <= word_min:
TypeError: '<=' not supported between instances of 'int' and 'str'

I made it into str('inf') - it did do the trick, so I assume is works the same :)

Comment (giladgressel) [reply python_fundamentals-master/08_file_io/08_01_words_analysis.py]: no they are not at all the same.  You are doing string comparison instead of integer comparison.
it will compare the ASCII values of the two stings and use that to evaluate <, > etc.

You got the error because you were trying to compare an integer with a string.  You should convert the other string (`word_min`) into an integer.  You probably need `len(word_min)` so you are comparing the lengths of the words.

File: python_fundamentals-master/07_classes_objects_methods/07_01_car.py (hunk 2 above)
Comment (giladgressel) [reply python_fundamentals-master/07_classes_objects_methods/07_01_car.py]: yes because you are overwriting the function name on line 19
The name of the method is `speed_up` but then you set on line 19
`self.speed_up = int` !! then will overwrite the method and now `speed_up` is an integer, not a method anymore.
I would guess it works the first time (line 29) but not the second time (line 30)
